lib/rpmsg/rpmsg_virtio.c

```diff
@@ -109,14 +108,17 @@ static int rpmsg_virtio_enqueue_buffer(struct rpmsg_virtio_device *rvdev,
 
 		/* Initialize buffer node */
 		vqbuf.buf = buffer;
-		vqbuf.len = len;
+		vqbuf.len = RPMSG_BUFFER_SIZE;
```

Comment:
you fixed the rpmsg buffer size here, but it is actually up to the allocator, it is not necessary to have fixed size buffer.

Reply:
> thanks for you patch, could you share your use case to have this feature?

We have several drivers use this feature:
1.rpmsg uart driver for console/gps/modem
2.rpmsg syslog driver for printf/log redirection
3.rpmsg netdev driver to talk with onchip mac layer(modem/wifi/ethernet)
4.rpmsg hostfs driver for file API redirection
5.rpmsg usrsock driver for sock API redirection
6.rpmsg clk driver for clk API redirection
Since these drivers:
1.Transfer the huge or variable length packet
2.The data transfer is the critical performance path
zero-copy could avoid one extra memory copy and then boost the performance.

> From your implementation, you add restrictions to the internal implementation, e.g. you use fixed buffer size, this will limit the future extension. And thus, want to know the need to bring back this feature.

Ok, I will refine the implementation to save the length in some reserved field as before.

Reply:
> you fixed the rpmsg buffer size here, but it is actually up to the allocator, it is not necessary to have fixed size buffer. You will need to also remember the buffer size.

Ok, I will save the length in some reserved field as before.

Reply:
@wjliang, after review the patch carefully again, I think the change is right for the current implementation.

> you fixed the rpmsg buffer size here, but it is actually up to the allocator,

Yes, rpmsg_virtio_shm_pool support the allocation with different size, but rpmsg master always allocate the fixed size(RPMSG_BUFFER_SIZE) now, so this limitation doesn't lead by this patch, but by the implementation. If master decide to change the buffer size in the future:
1.Negotiate individual buffer size dynamically(https://github.com/OpenAMP/open-amp/pull/155), we can update as:
  vqbuf.len = rvdev->config.rxbuf_size;
2.Support the variable buffer length, we can:
   a.Reserve 32bit space before the buffer, and save the length here, or
   b.Add a new length array, save the length into it by buffer index
We even extend the len of rpmsg_get_tx_payload_buffer as in/out parameter, so the caller can pass in the required size.

> it is not necessary to have fixed size buffer. You will need to also remember the buffer size.

As mention before, this patch is compatible with the current implementation but still easily modify to cover the extension.

Reply:
> Yes, rpmsg_virtio_shm_pool support the allocation with different size, but rpmsg master always allocate the fixed size(RPMSG_BUFFER_SIZE) now, so this limitation doesn't lead by this patch, but by the implementation. If master decide to change the buffer size in the future:
> 1.Negotiate individual buffer size dynamically(#155), we can update as:
> vqbuf.len = rvdev->config.rxbuf_size;

I think the best is the buffer size is allocated based on the need of message/data, it can be, have default size, if the data is larger, the allocator can decide to enlarge the buffer. I prefer not to add more limitation to the existing implementation.
Even this feature has to assume the buffer size, it is better to fix the size in the new functions rather than change the existing function. As we can consider to conditionally enable the new feature.

Reply:
Ok, please the review the new patch I just send out, which don't add more limitation to the existing implemenation.